packages/react-components/react-table/src/hooks/selectionManager.ts

```diff
@@ -0,0 +1,112 @@
+import { SelectionMode } from './types';
+
+type OnSelectionChangeCallback = (selectedItems: Set<SelectionItemId>) => void;
+
+export interface SelectionManager {
+  toggleSelect(id: SelectionItemId): void;
+  select(id: SelectionItemId): void;
+  deSelect(id: SelectionItemId): void;
+  clearSelection(): void;
+  isSelected(id: SelectionItemId): boolean;
+  toggleSelectAll(itemIds: SelectionItemId[]): void;
+}
+
+export type SelectionItemId = string | number;
+
+export function createSelectionManager(
+  mode: SelectionMode,
+  onSelectionChange: OnSelectionChangeCallback = () => undefined,
+): SelectionManager {
+  const selectedItems = new Set<SelectionItemId>();
+  return mode === 'multiselect'
+    ? createMultipleSelectionManager(selectedItems, onSelectionChange)
+    : createSingleSelectionManager(selectedItems, onSelectionChange);
+}
+
+function createMultipleSelectionManager(
+  set: Set<SelectionItemId>,
+  onSelectionChange: OnSelectionChangeCallback,
+): SelectionManager {
+  const toggleSelectAll = (itemIds: SelectionItemId[]) => {
+    if (set.size === itemIds.length) {
+      set.clear();
+    } else {
+      itemIds.forEach(itemId => set.add(itemId));
+    }
+
+    onSelectionChange(new Set(set));
+  };
+
+  const toggleSelect = (itemId: SelectionItemId) => {
+    if (set.has(itemId)) {
+      set.delete(itemId);
+    } else {
+      set.add(itemId);
+    }
+
+    onSelectionChange(new Set(set));
+  };
+
+  const select = (itemId: SelectionItemId) => {
+    set.add(itemId);
+    onSelectionChange(new Set(set));
+  };
+
+  const deSelect = (itemId: SelectionItemId) => {
+    set.delete(itemId);
+    onSelectionChange(new Set(set));
+  };
+
+  const clearSelection = () => {
+    set.clear();
+    onSelectionChange(new Set(set));
+  };
+
+  const isSelected = (itemId: SelectionItemId) => {
+    return set.has(itemId);
+  };
+
+  return {
+    toggleSelect,
+    select,
+    deSelect,
+    clearSelection,
+    isSelected,
+    toggleSelectAll,
+  };
+}
+
+function createSingleSelectionManager(
+  set: Set<SelectionItemId>,
+  onSelectionChange: OnSelectionChangeCallback,
+): SelectionManager {
```

Comment:
While it makes sense to use `Set` in case of multiple - there it has no sense 🙃

What do you think about this refactor? ⬇️

```ts
function createSingleSelectionManager(
  onSelectionChange: OnSelectionChangeCallback,
): SelectionManager {
  let selectedItemId: SelectionItemId | null = null;
  
  const toggleSelect = (itemId: SelectionItemId) => {
    selectedItemId = isSelected(itemId) ? null : itemId
    onSelectionChange(new Set([itemId]));
  };

  const clearSelection = () => {
    selectedItemId = null
    onSelectionChange(new Set([]));
  };

  const isSelected = (itemId: SelectionItemId) => {
    return selectedItemId === itemId
  };

  const select = (itemId: SelectionItemId) => {
    selectedItemId = itemId
    onSelectionChange(new Set([itemId]));
  };

  return {
    deSelect: clearSelection,
    select,
    toggleSelectAll: () => undefined,
    toggleSelect,
    clearSelection,
    isSelected,
  };
}
```